### ISSUE TITLE
vinyl: don't use tx_index from coeio thread

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#1603]
tx_index should only be used from TX thread.
Please refactor vy_read_task() and remove useless mutexes.

**Comment by rtsisyk** [comment]
[gh-1603.diff.txt](https://github.com/tarantool/tarantool/files/403652/gh-1603.diff.txt)

```
vinyl/hermitage.test.lua                                        [ fail ]

Test failed! Result content mismatch:
--- vinyl/hermitage.result      Thu Aug  4 14:48:24 2016
+++ vinyl/hermitage.reject      Fri Aug  5 14:25:54 2016
@@ -278,7 +278,7 @@
 ...
 c2("t:get{1}") -- {1, 10}
 ---
-- - [1, 10]
+- - [1, 11]
 ...
 c2:commit() -- ok
 ---
@@ -459,7 +459,7 @@
 ...
 c1("t:get{3}") -- nothing
 ---
-- 
+- - [3, 30]
 ...
 c1:commit() -- ok
 ---
@@ -518,7 +518,7 @@
 ...
 c2("t:get{1}") -- {1, 10}
 ---
-- - [1, 10]
+- - [1, 20]
 ...
 c2:commit() -- rollback -- conflict
 ---
@@ -636,7 +636,7 @@
 ...
 c1("t:get{2}") -- {2, 20}
 ---
-- - [2, 20]
+- - [2, 18]
 ...
 c1:commit() -- ok
 ---
@@ -699,7 +699,7 @@
 ...
 c1("t:get{2}") -- finds nothing
 ---
-- 
+- - [2, 18]
 ...
 c1:commit() -- rollback
 ---

Last 15 lines of Tarantool Log file:
vinyl/info.test.lua                                             [ pass ]
vinyl/large.test.lua                                            [ disabled ]
vinyl/mvcc.test.lua                                             [ fail ]

Test failed! Result content mismatch:
--- vinyl/mvcc.result   Fri Aug  5 13:45:52 2016
+++ vinyl/mvcc.reject   Fri Aug  5 14:25:55 2016
@@ -542,7 +542,7 @@
 ...
 c1("t:get{1}")
 ---
-- - [1]
+- - [1, 'c2']
 ...
 c2:commit()
 ---
@@ -553,7 +553,7 @@
 --
 c1("t:get{1}")
 ---
-- - [1]
+- - [1, 'c2']
 ...
 -- commits ok since is a read only transaction
 c1:commit()
@@ -2684,7 +2684,7 @@
 --
 c2("t:get{1}") -- finds nothing
 ---
-- 
+- - [1, 10]
 ...
 --
 c2("t:replace{1, 15}")
@@ -3035,7 +3035,7 @@
 --
 c2("t:get{1}") -- {1, 3}
 ---
-- - [1, 3]
+- - [1, 12]
 ...
 --
 c4("t:get{1}") -- {1, 12}
@@ -3045,7 +3045,7 @@
 --
 c5("t:get{1}") -- {1, 9}
 ---
-- - [1, 9]
+- - [1, 12]
 ...
 --
 c6("t:get{1}") -- {1, 12}
@@ -3068,12 +3068,12 @@
 --
 c1("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c7("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c2:rollback()
@@ -3088,7 +3088,7 @@
 --
 c5("t:get{1}") -- {1, 9}
 ---
-- - [1, 9]
+- - [1, 12]
 ...
 --
 c6("t:get{1}") -- {1, 12}
@@ -3111,12 +3111,12 @@
 --
 c1("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c7("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c4:rollback()
@@ -3126,7 +3126,7 @@
 --
 c5("t:get{1}") -- {1, 9}
 ---
-- - [1, 9]
+- - [1, 12]
 ...
 --
 c6("t:get{1}") -- {1, 12}
@@ -3149,12 +3149,12 @@
 --
 c1("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c7("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c5:rollback()
@@ -3182,12 +3182,12 @@
 --
 c1("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c7("t:get{1}") -- nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c6:rollback()
@@ -3210,12 +3210,12 @@
 --
 c1("t:get{1}") -- finds nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c7("t:get{1}") -- finds nothing
 ---
-- 
+- - [1, 12]
 ...
 --
 c1:rollback()

```

**Comment by rtsisyk** [comment]
Fixed by 66f2df684cbd5c6312a851f9395b0d7b9b00b2e0
